torch/jit.py

```diff
@@ -1,26 +1,496 @@
+        # we mix with from accidentally scrambling us
+        #
+        # NB: Class variables are also accessible via self!
+        kwargs["time"] = time  # also want to pass onto ktrace
+        cls.__ktrace_kwargs = kwargs
+        cls.__enabled = enabled
+        cls.__time = time
+        cls.__model_name = name
+
+        # Monkey patch the constructor and forward functions *inplace*
+        cls.__old_forward = cls.forward
+        cls.forward = cls.__new_forward
+        cls.__old_init = cls.__init__
+        cls.__init__ = cls.__new_init
+
+    def __new_init(self, *args, **kwargs):
+        try:
+            # __old_init is assumed to handle super call
+            self.__old_init(*args, **kwargs)
+        except TypeError as e:
+            # If this fails here, the user probably didn't use this as a class
+            # decorator
+            if "super" in str(e):
+                raise_from(TypeError("torch.jit.compile must be used as a class decorator; "
+                                     "using it on an already defined class is not valid."
+                                     "\n\nOriginal error: {}".format(str(e))), e)
+            else:
+                raise
+        model_name = self.__model_name if self.__model_name else type(self).__name__
+        self.__name = "jit_{}_{}".format(model_name, _CompiledMixin.__next_id)
+        _CompiledMixin.__next_id += 1
+        self.__ktrace_cache = {}
+        self.__next_ktrace_id = 0
+
+    def __process_args(self, args):
+        in_vars, in_struct = _flatten(args, self.state_dict(keep_vars=True).values())
+        is_volatile, in_vars_key = vars_key(in_vars)
+        in_key = (in_vars_key, in_struct)
+        return in_vars, in_struct, is_volatile, in_key
+
+    # NB: In principle, there could also be a 'raw' version of this compiler,
+    # but since the logic is so complicated, testing code wouldn't benefit much
+    def __new_forward(self, *args):
+        if _JIT_DISABLE or not self.__enabled:
+            with _time(self.__name, "unoptimized", self.__time):
+                # Call to the saved old forward function
+                return self.__old_forward(*args)
+        in_vars, in_struct, is_volatile, in_key = self.__process_args(args)
+        ktrace = self.__ktrace_cache.get(in_key)
+        if ktrace is None:
+            ktrace_name = '{}_{}'.format(self.__name, self.__next_ktrace_id)
+            self.__next_ktrace_id += 1
+            ktrace = TraceForKey(ktrace_name, in_key, volatile=is_volatile, **self.__ktrace_kwargs)
+            self.__ktrace_cache[in_key] = ktrace
+        closure = ktrace.maybe_closure()
+        if closure is not None:
+            # We already compiled it!  Run it directly, and
+            # use the saved out_struct to unflatten.
+            with _time(ktrace.name, "optimized", self.__time):
+                out_vars = closure()(*in_vars)
+                out_struct = ktrace.out_struct
+        else:
```

Comment:
Is it possible to check here that the cache already has an active (but incomplete) trace for this configuration, and just skip adding another trace such that TraceForKey only needs 1 trace?

Reply:
CC @apaszke. I think this might be possible (but fiddly): basically, we'd execute the saved incomplete trace, and then some how fixup the `grad_fn` of the returned Variables to point to the old grad_fn graph (from the first trace) which knows how to feed into backwards. Maybe Adam knows why we didn't do this in the first place.

Reply:
It wouldn't work in all cases. We need to keep multiple traces alive